File: androidApp/src/main/java/com/app/pustakam/android/fileUtils/FileOps.kt
package com.app.pustakam.android.fileUtils

import android.app.Activity
import android.content.ContentValues
import android.content.Context
import android.graphics.Bitmap
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.provider.MediaStore
import com.app.pustakam.data.models.response.notes.NoteContentModel
import com.app.pustakam.util.ContentType
import java.io.File
import java.io.FileOutputStream
import java.io.IOException


fun createFileWithFolders(context: Activity, folderPath : String, fileName : String) : File {
    var file : File = File("")
    try{
        val baseDir = File(context.filesDir,folderPath)
        if (!baseDir.exists()) {
            if (!baseDir.mkdirs()) {
            throw IOException("Failed to create directories: ${baseDir.absolutePath}")
            }
        }
        file = File(baseDir.absolutePath, fileName)
        if (!file.exists()) {
            try {
                file.createNewFile()
            } catch (e: IOException) {
                throw IOException("Failed to create file: ${file.absolutePath}", e)
            }
        }
    } catch (e: Exception){
        e.printStackTrace()
    }
    return file
}

fun deleteFile(filePath : String){
    val file = File(filePath)
    if (file.exists() && file.isFile) {
        file.delete()
    }
}

// 🔧 15-Jul-2026 Phase 2.3: lazy thumbnails — a small JPEG (max ~512px, q70) written next to the
//   app's files under thumbnails/, so lists and placeholders never decode the full image or a
//   video stream. Returns the thumbnail's absolute path, or null when generation fails (callers
//   simply keep the current fallback behavior).
//   • IMAGE/GIF: bounds-decode + inSampleSize (no full-size bitmap in memory)
//   • VIDEO:     first frame at ~1s via MediaMetadataRetriever, scaled down
//   Usage: generateThumbnail(context, media.localPath!!, media.type)
private const val THUMBNAIL_MAX_DIMENSION = 512

fun generateThumbnail(context: Context, sourcePath: String, contentType: ContentType): String? {
    return try {
        val source = File(sourcePath)
        if (!source.exists()) return null
        val bitmap: Bitmap = when (contentType) {
            ContentType.IMAGE, ContentType.GIF -> decodeDownsampled(sourcePath)
            ContentType.VIDEO -> {
                val retriever = android.media.MediaMetadataRetriever()
                try {
                    retriever.setDataSource(sourcePath)
                    val frame = retriever.getFrameAtTime(1_000_000) // ~1s in — skips black lead-ins
                        ?: retriever.getFrameAtTime(0)
                    frame?.let { scaleDown(it) }
                } finally {
                    retriever.release()
                }
            }
            else -> null
        } ?: return null
        val thumbDir = File(context.filesDir, "thumbnails").apply { mkdirs() }
        val thumbFile = File(thumbDir, "${source.nameWithoutExtension}_thumb.jpg")
        FileOutputStream(thumbFile).use { out ->
            bitmap.compress(Bitmap.CompressFormat.JPEG, 70, out)
        }
        thumbFile.absolutePath
    } catch (e: Exception) {
        e.printStackTrace()
        null
    }
}

// 🔧 15-Jul-2026 Phase 2.3: decode at reduced resolution straight from disk.
private fun decodeDownsampled(path: String): Bitmap? {
    val bounds = android.graphics.BitmapFactory.Options().apply { inJustDecodeBounds = true }
    android.graphics.BitmapFactory.decodeFile(path, bounds)
    if (bounds.outWidth <= 0 || bounds.outHeight <= 0) return null
    var sampleSize = 1
    while (bounds.outWidth / (sampleSize * 2) >= THUMBNAIL_MAX_DIMENSION ||
        bounds.outHeight / (sampleSize * 2) >= THUMBNAIL_MAX_DIMENSION
    ) sampleSize *= 2
    val opts = android.graphics.BitmapFactory.Options().apply { inSampleSize = sampleSize }
    return android.graphics.BitmapFactory.decodeFile(path, opts)
}

// 🔧 15-Jul-2026 Phase 2.3: proportional scale so the longest side is THUMBNAIL_MAX_DIMENSION.
private fun scaleDown(bitmap: Bitmap): Bitmap {
    val longest = maxOf(bitmap.width, bitmap.height)
    if (longest <= THUMBNAIL_MAX_DIMENSION) return bitmap
    val scale = THUMBNAIL_MAX_DIMENSION.toFloat() / longest
    return Bitmap.createScaledBitmap(
        bitmap, (bitmap.width * scale).toInt().coerceAtLeast(1),
        (bitmap.height * scale).toInt().coerceAtLeast(1), true
    )
}
fun saveBitmapToFile(bitmap: Bitmap, filePath: String): Boolean {
    val file = File(filePath)
   return saveBitmapToFile(bitmap,file)
}
fun saveBitmapToFile(bitmap: Bitmap,file: File): Boolean{
    return try {
        file.parentFile?.mkdirs() // Ensure the directory exists
        FileOutputStream(file).use { out ->
            bitmap.compress(Bitmap.CompressFormat.PNG,100 , out) // Save as PNG with 100% quality
        }
        true // Success
    } catch (e: IOException) {
        e.printStackTrace()
        false // Failure
    }
}

// 🔧 14-Jul-2026: NEW FEATURE — "Save media to device".
//   Images/Videos are written silently into the system Gallery (MediaStore);
//   Audio/PDF/DOCX/GIF are exported through the Storage-Access-Framework picker
//   (see MediaSaveOverlay in NotesEditorView.kt) into a user-chosen folder
//   defaulting to Downloads. Everything below is additive — no existing API changed.

// 🔧 14-Jul-2026: MIME type for a media block (used by MediaStore + the SAF picker).
//   Usage: mimeTypeFor(media.type)  ->  "image/png", "video/mp4", "audio/mpeg", ...
fun mimeTypeFor(type: ContentType): String = when (type) {
    ContentType.IMAGE -> "image/png"
    ContentType.VIDEO -> "video/mp4"
    ContentType.AUDIO -> "audio/mpeg"
    ContentType.PDF -> "application/pdf"
    ContentType.DOCX -> "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
    ContentType.GIF -> "image/gif"
    else -> "application/octet-stream"
}

// 🔧 14-Jul-2026: A human/file-system friendly name for the exported/saved file.
//   Falls back to the source file name, then to a timestamped default.
//   Usage: suggestedFileName(media)  ->  "Video-2.mp4"
fun suggestedFileName(media: NoteContentModel.MediaContent): String {
    val ext = media.type.getExt().ifEmpty { "" }
    val fromSource = media.localPath?.substringAfterLast('/')?.takeIf { it.isNotBlank() }
    val base = media.title.takeIf { it.isNotBlank() }
        ?: fromSource?.substringBeforeLast('.')
        ?: "Pustakam-${System.currentTimeMillis()}"
    // If we already have a real file name with an extension, keep it as-is.
    if (fromSource != null && fromSource.contains('.')) return fromSource
    return if (ext.isNotEmpty()) "$base$ext" else base
}

// 🔧 14-Jul-2026: Save an IMAGE or VIDEO straight into the device Gallery.
//   Uses scoped-storage MediaStore on Android 10+ (no runtime permission needed);
//   on older devices it relies on WRITE_EXTERNAL_STORAGE (declared in the manifest,
//   maxSdkVersion=28). Returns true on success.
//   Usage: val ok = saveMediaToGallery(context, media)
fun saveMediaToGallery(context: Context, media: NoteContentModel.MediaContent): Boolean {
    val sourcePath = media.localPath?.takeIf { it.isNotEmpty() } ?: media.url.takeIf { it.isNotEmpty() }
    if (sourcePath.isNullOrEmpty()) return false
    val source = File(sourcePath)
    if (!source.exists()) return false

    val resolver = context.contentResolver
    val fileName = suggestedFileName(media)
    val mime = mimeTypeFor(media.type)
    val isVideo = media.type == ContentType.VIDEO

    // Collection + relative sub-folder differ for images vs videos.
    val collection = if (isVideo)
        MediaStore.Video.Media.EXTERNAL_CONTENT_URI else MediaStore.Images.Media.EXTERNAL_CONTENT_URI
    val relativeDir = if (isVideo)
        "${Environment.DIRECTORY_MOVIES}/Pustakam" else "${Environment.DIRECTORY_PICTURES}/Pustakam"

    val values = ContentValues().apply {
        put(MediaStore.MediaColumns.DISPLAY_NAME, fileName)
        put(MediaStore.MediaColumns.MIME_TYPE, mime)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            put(MediaStore.MediaColumns.RELATIVE_PATH, relativeDir)
            put(MediaStore.MediaColumns.IS_PENDING, 1) // hide until fully written
        }
    }

    return try {
        val uri: Uri = resolver.insert(collection, values) ?: return false
        resolver.openOutputStream(uri)?.use { out ->
            source.inputStream().use { input -> input.copyTo(out) }
        } ?: return false
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            values.clear()
            values.put(MediaStore.MediaColumns.IS_PENDING, 0) // publish
            resolver.update(uri, values, null, null)
        }
        true
    } catch (e: Exception) {
        e.printStackTrace()
        false
    }
}

// 🔧 14-Jul-2026: Copy a media file's bytes into a SAF-picked destination Uri.
//   Used for AUDIO/PDF/DOCX/GIF once the user picks a location via the
//   ACTION_CREATE_DOCUMENT picker. Returns true on success.
//   Usage: writeMediaToUri(context, media, pickedUri)
fun writeMediaToUri(context: Context, media: NoteContentModel.MediaContent, destination: Uri): Boolean {
    val sourcePath = media.localPath?.takeIf { it.isNotEmpty() } ?: media.url.takeIf { it.isNotEmpty() }
    if (sourcePath.isNullOrEmpty()) return false
    val source = File(sourcePath)
    if (!source.exists()) return false
    return try {
        context.contentResolver.openOutputStream(destination)?.use { out ->
            source.inputStream().use { input -> input.copyTo(out) }
        } ?: return false
        true
    } catch (e: Exception) {
        e.printStackTrace()
        false
    }
}

File: iosApp/iosApp/file/FileOps.swift

import Foundation
import Photos          // 🔧 14-Jul-2026: save images/videos to the Photos gallery
import UIKit           // 🔧 14-Jul-2026: UIAlertController action sheet for the save-options chooser
import shared          // 🔧 14-Jul-2026: NoteContentModel / ContentType for the save dispatcher


//Creating A file in specific path
func createFilepath(in folderName: String, fileName: String, contents: Data) -> URL? {
    guard let folderURL = createFolder(named: folderName) else { return nil }
    let fileURL = folderURL.appendingPathComponent(fileName)

    FileManager.default.createFile(atPath: fileURL.path, contents: contents, attributes: nil)
    return fileURL
}

//Create a New Folder in Documents Directory

func createFolder(named folderName: String) -> URL? {
    let documentsDirectory = getDocumentsDirectory()
    let folderURL = documentsDirectory.appendingPathComponent(folderName)

    if !FileManager.default.fileExists(atPath: folderURL.path) {
        do {
            try FileManager.default.createDirectory(at: folderURL, withIntermediateDirectories: true, attributes: nil)
            print("Folder created at: \(folderURL.path)")
        } catch {
            print("Error creating folder: \(error)")
            return nil
        }
    }

    return folderURL
}


//Get a Specific Directory Path (e.g., Documents)

func getDocumentsDirectory() -> URL {
    FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
}


func getATempFilePath(extension ext : String) -> URL{
    FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString + ext)
}

//save data into location
func saveData(url : URL, data : Data){
    do {
        try data.write(to: url)
    }catch {
        print("❌ Failed to save: \(error)")
    }
}

//move data into location
func moveData(from url  : URL , to destinationUrl  : URL){
    do{
        try FileManager.default.copyItem(at: url, to: destinationUrl)
    }catch {
        print("❌ Failed to move: \(error)")
    }
}
// save media file
func saveImageFile(data content : Data, in folder : String,to filePath : String ,) -> String {
    
    guard let  fileSavingPath = createFilepath(in: folder, fileName: filePath, contents: content) else {return ""}
    saveData(url: fileSavingPath , data: content)
    return fileSavingPath.path
}

func saveVideoFile(in folder : String,to filePath : String, url : URL) -> String{
     var localpath  = ""
    if let videoData = try? Data(contentsOf: url) {
       let   savedURL = createFilepath(in: folder, fileName: filePath, contents: videoData)
        print("Saved at: \(String(describing: savedURL))")
    } else {
        print("Failed to read video data.")
    }
    
    return localpath
}
// 🔧 14-Jul-2026: NEW — delete a media file from disk when its note-content block is removed
//   (Android parity: NoteEditorViewModel.removeContent deletes the local file too).
//   Usage: deleteFile(filePath: media.localPath) — safe to call if the file no longer exists.
func deleteFile(filePath: String) {
    guard !filePath.isEmpty, FileManager.default.fileExists(atPath: filePath) else { return }
    do {
        try FileManager.default.removeItem(atPath: filePath)
    } catch {
        print("❌ Failed to delete file: \(error)")
    }
}

// 🔧 14-Jul-2026: NEW FEATURE — "Save media to device".
//   IMAGE/VIDEO are saved silently into the Photos gallery; AUDIO/PDF/DOCX/GIF are
//   exported through the Files document picker (default = Documents) so the user
//   picks the destination. Everything below is additive.

// 🔧 14-Jul-2026: Entry point called by the media cards' save button.
//   Usage:  saveMediaToDevice(media: contentImage)
func saveMediaToDevice(media: NoteContentModel.MediaContent) {
    let path = media.getMediaUrl()
    guard !path.isEmpty, FileManager.default.fileExists(atPath: path) else {
        print("❌ Save failed — source file missing at \(path)")
        return
    }
    let fileURL = URL(fileURLWithPath: path)
    switch media.type {
    case .image, .video:
        // 🔧 14-Jul-2026: CHANGED — was a silent gallery save; now shows a bottom action sheet
        //   with two options so the user picks Gallery vs. a file location.
        presentSaveOptions(fileURL: fileURL, isVideo: media.type == .video)
    default:
        // audio / pdf / docx / gif can't live in Photos → let the user pick a folder.
        presentDocumentExporter(fileURL: fileURL)
    }
}

// 🔧 14-Jul-2026: NEW FEATURE — bottom sheet with the two save choices for IMAGE/VIDEO.
//   Option 1 "Save to Gallery"          → saveMediaToGallery (silent Photos write).
//   Option 2 "Save to location as file" → presentDocumentExporter (Files picker, default Documents).
//   Uses a UIAlertController action sheet so it works from a free function (no View change needed).
//   Usage:  presentSaveOptions(fileURL: url, isVideo: true)
func presentSaveOptions(fileURL: URL, isVideo: Bool) {
    guard let top = topMostViewController() else {
        print("❌ No view controller available to present the save options")
        return
    }
    let sheet = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
    sheet.addAction(UIAlertAction(title: "Save to Gallery", style: .default) { _ in
        saveMediaToGallery(fileURL: fileURL, isVideo: isVideo)
    })
    sheet.addAction(UIAlertAction(title: "Save to location as file", style: .default) { _ in
        presentDocumentExporter(fileURL: fileURL)
    })
    sheet.addAction(UIAlertAction(title: "Cancel", style: .cancel))
    // iPad requires an anchor for action sheets.
    if let popover = sheet.popoverPresentationController {
        popover.sourceView = top.view
        popover.sourceRect = CGRect(x: top.view.bounds.midX, y: top.view.bounds.maxY, width: 0, height: 0)
        popover.permittedArrowDirections = []
    }
    top.present(sheet, animated: true)
}

// 🔧 14-Jul-2026: Save an image or video file into the Photos library.
//   Requests add-only access (NSPhotoLibraryAddUsageDescription is declared in Info.plist).
//   Usage:  saveMediaToGallery(fileURL: url, isVideo: true)
func saveMediaToGallery(fileURL: URL, isVideo: Bool) {
    PHPhotoLibrary.requestAuthorization(for: .addOnly) { status in
        guard status == .authorized || status == .limited else {
            print("❌ Photo library permission denied")
            return
        }
        PHPhotoLibrary.shared().performChanges({
            if isVideo {
                _ = PHAssetCreationRequest.creationRequestForAssetFromVideo(atFileURL: fileURL)
            } else {
                _ = PHAssetCreationRequest.creationRequestForAssetFromImage(atFileURL: fileURL)
            }
        }) { success, error in
            if success {
                print("✅ Saved to Photos gallery")
            } else {
                print("❌ Save to Photos failed: \(String(describing: error))")
            }
        }
    }
}

// for large files
func copyFile(to folderName: String, fileName: String, from sourceURL: URL) -> URL? {
    guard let folderURL = createFolder(named: folderName) else { return nil }
    let destinationURL = folderURL.appendingPathComponent(fileName)

    do {
        try FileManager.default.copyItem(at: sourceURL, to: destinationURL)
        return destinationURL
    } catch {
        print("Failed to copy file: \(error)")
        return nil
    }
}

// 🔧 15-Jul-2026 iOS parity (Phase 2.3): lazy thumbnails — a small JPEG (max ~512px, q0.7) written
//   under Documents/thumbnails/, so list cards never decode the full image or open a video stream.
//   Android counterpart: fileUtils/FileOps.kt generateThumbnail(). Returns the thumbnail's path,
//   or nil on failure (callers keep their current fallback behavior).
//   • IMAGE/GIF: UIImage downscale via UIGraphicsImageRenderer
//   • VIDEO:     frame at ~1s via AVAssetImageGenerator (skips black lead-ins), fallback frame 0
//   Usage: generateThumbnail(sourcePath: media.localPath!, type: media.type)
import AVFoundation

// 🔧 30-Jul-2026 Phase 4 — size and quality come from the SHARED ThumbnailPolicy so iOS and Android
//   cannot drift. Accessor functions, not the `const val`s: a const inside a Kotlin object has no
//   guaranteed ObjC/Swift export shape. The DECODING below stays native (UIImage / AVAssetImageGenerator).
private let thumbnailMaxDimension = CGFloat(ThumbnailPolicy.shared.maxDimensionPx())
private let thumbnailJpegQuality = CGFloat(ThumbnailPolicy.shared.jpegQuality()) / 100.0

func generateThumbnail(sourcePath: String, type: ContentType) -> String? {
    let sourceURL = URL(fileURLWithPath: sourcePath)
    guard FileManager.default.fileExists(atPath: sourcePath) else { return nil }

    var image: UIImage?
    if type == ContentType.image || type == ContentType.gif {
        image = UIImage(contentsOfFile: sourcePath)
    } else if type == ContentType.video {
        let generator = AVAssetImageGenerator(asset: AVURLAsset(url: sourceURL))
        generator.appliesPreferredTrackTransform = true
        // 🔧 30-Jul-2026 Phase 4 — frame offset from the shared ThumbnailPolicy (µs → seconds)
        let frameSeconds = Double(ThumbnailPolicy.shared.videoFrameMicros()) / 1_000_000.0
        if let cg = try? generator.copyCGImage(at: CMTime(seconds: frameSeconds, preferredTimescale: 600), actualTime: nil) {
            image = UIImage(cgImage: cg)
        } else if let cg = try? generator.copyCGImage(at: .zero, actualTime: nil) {
            image = UIImage(cgImage: cg)
        }
    }
    guard let source = image else { return nil }

    // proportional downscale so the longest side is thumbnailMaxDimension
    let longest = max(source.size.width, source.size.height)
    let scale = longest > thumbnailMaxDimension ? thumbnailMaxDimension / longest : 1
    let targetSize = CGSize(width: max(source.size.width * scale, 1),
                            height: max(source.size.height * scale, 1))
    let scaled = UIGraphicsImageRenderer(size: targetSize).image { _ in
        source.draw(in: CGRect(origin: .zero, size: targetSize))
    }
    guard let data = scaled.jpegData(compressionQuality: thumbnailJpegQuality),
          let folderURL = createFolder(named: "thumbnails") else { return nil }
    let thumbURL = folderURL.appendingPathComponent(
        sourceURL.deletingPathExtension().lastPathComponent + "_thumb.jpg")
    do {
        try data.write(to: thumbURL)
        return thumbURL.path
    } catch {
        print("Failed to write thumbnail: \(error)")
        return nil
    }
}
